File: src/main/java/com/github/kory33/signvote/listeners/VotePointProtector.java
package com.github.kory33.signvote.listeners;

import com.github.kory33.signvote.core.SignVote;
import com.github.kory33.signvote.utils.BlockUtil;
import org.bukkit.block.Block;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.material.MaterialData;

import java.util.Set;

/**
 * A listener class that aims to protect blocks behind vote-points
 */
public class VotePointProtector implements Listener {
    private final SignVote plugin;

    public VotePointProtector(SignVote plugin) {
        this.plugin = plugin;
        plugin.getServer().getPluginManager().registerEvents(this, plugin);
    }

    @EventHandler
    public void onVotePointBaseBlockBreak(BlockBreakEvent event) {
        Block brokenBlock = event.getBlock();

        Set<Block> attachedVotePoints = BlockUtil.getBlocksAdjacentTo(brokenBlock);
        attachedVotePoints.removeIf(block -> {
            // ignore if the block is not a sign
            MaterialData state = block.getState().getData();
            if (!(state instanceof org.bukkit.material.Sign)) {
                return true;
            }

            // ignore if the sign is not attached to the broken block
            org.bukkit.material.Sign signMaterial = (org.bukkit.material.Sign) state;
            Block attachedBlock = block.getRelative(signMaterial.getAttachedFace());
            if (!attachedBlock.equals(brokenBlock)) {
                return true;
            }

            // ignore if the sign is not a SignVote's vote-point
            return !this.plugin.getAPI().isSignVoteSign(block);
        });

        if (attachedVotePoints.isEmpty()) {
            return;
        }

        event.setCancelled(true);
    }
}

File: src/main/kotlin/com/github/kory33/signvote/listeners/VotePointProtector.kt
package com.github.kory33.signvote.listeners

import com.github.kory33.signvote.core.SignVote
import com.github.kory33.signvote.utils.BlockUtil
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.block.BlockBreakEvent

/**
 * A listener class that aims to protect blocks behind vote-points
 */
class VotePointProtector(private val plugin: SignVote) : Listener {

    init {
        plugin.server.pluginManager.registerEvents(this, plugin)
    }

    @EventHandler
    fun onVotePointBaseBlockBreak(event: BlockBreakEvent) {
        val brokenBlock = event.block

        val protectTarget = BlockUtil.getBlocksAdjacentTo(brokenBlock)
                .firstOrNull { block ->
                    val signMaterial = block.state.data as? org.bukkit.material.Sign ?: return@firstOrNull false

                    // ignore if the sign is not attached to the broken block
                    val attachedBlock = block.getRelative(signMaterial.attachedFace)
                    if (attachedBlock != brokenBlock) {
                        return@firstOrNull false
                    }

                    return@firstOrNull this.plugin.api!!.isSignVoteSign(block)
                }

        if(protectTarget != null) {
            event.isCancelled = true
        }
    }
}
